### PR TITLE
databricks-connect: init at 7.1.0

### PR DESCRIPTION
**Pull request by kfollesdal**
Python package [databricks-connect](https://pypi.org/project/databricks-connect/) is missing in nixpkgs. Used this as an opportunity to try make my first nix expression and nixpkgs. 

Very happy for reviews and questions. 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] Ubuntu
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x]  Tested execution of **relevant** binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: lib/licenses.nix [hunk 1]
```diff
@@ -436,6 +436,12 @@ lib.mapAttrs (n: v: v // { shortName = n; }) {
   };
 
   # Proprietary binaries; free to redistribute without modification.
+  databricks = {
+    fullName = "Databricks Proprietary License";
+    url = "https://pypi.org/project/databricks-connect";
+
```
Comment (catern) [inline lib/licenses.nix:442]: Shouldn't you mark free = false here?

File: pkgs/development/python-modules/databricks-connect/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  disabled = !isPy37;
+
+  src = fetchPypi { inherit pname version sha256; };
+
+  sourceRoot = ".";
+
+  propagatedBuildInputs = [ py4j six pkgs.jdk ];
+
+  doCheck = false;
```
Comment (catern) [inline pkgs/development/python-modules/databricks-connect/default.nix:16]: Can you try and run the tests, or explain why they don't work?

File: pkgs/development/python-modules/databricks-connect/default.nix [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+buildPythonPackage rec {
+  pname = "databricks-connect";
+  version = "7.1.0";
+  sha256 = "996a9d0f271f6c7edbd2d85b2efb6ff4e58d15222e80f87ca17fdbf224e17056";
+
+  disabled = !isPy37;
+
+  src = fetchPypi { inherit pname version sha256; };
+
+  sourceRoot = ".";
```
Comment (catern) [inline pkgs/development/python-modules/databricks-connect/default.nix:12]: What's the motivation for this exactly?

File: pkgs/development/python-modules/databricks-connect/default.nix [hunk 4]
```diff
@@ -0,0 +1,31 @@
+{ lib, isPy37, pkgs, buildPythonPackage, fetchPypi, six, py4j }:
+
+buildPythonPackage rec {
+  pname = "databricks-connect";
+  version = "7.1.0";
+  sha256 = "996a9d0f271f6c7edbd2d85b2efb6ff4e58d15222e80f87ca17fdbf224e17056";
```
Comment (catern) [inline pkgs/development/python-modules/databricks-connect/default.nix:6]: The pname and version are fine here, but the sha256 would be more normal in the args to fetchPypi

File: pkgs/development/python-modules/databricks-connect/default.nix [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+  disabled = !isPy37;
```
Comment (catern) [inline pkgs/development/python-modules/databricks-connect/default.nix:8]: From looking on pypi, it looks like this package supports a pretty wide range of Python versions - do you even need this disabled flag? It actually seems to support every version in Nixpkgs - though if you want to limit it to isPy3, that would be reasonable too.

File: pkgs/development/python-modules/databricks-connect/default.nix (hunk 2 above)
Comment (kfollesdal) [reply pkgs/development/python-modules/databricks-connect/default.nix:16]: When running test get the following error. Have to look more in to it. 
```
running test
WARNING: Testing via this command is deprecated and will be removed in a future version. Users looking for a generic test entry point independent of test runner are encouraged to use tox.
running egg_info
writing databricks_connect.egg-info/PKG-INFO
writing dependency_links to databricks_connect.egg-info/dependency_links.txt
writing entry points to databricks_connect.egg-info/entry_points.txt
writing requirements to databricks_connect.egg-info/requires.txt
writing top-level names to databricks_connect.egg-info/top_level.txt
package init file 'deps/bin/__init__.py' not found (or not a regular file)
package init file 'deps/sbin/__init__.py' not found (or not a regular file)
package init file 'deps/jars/__init__.py' not found (or not a regular file)
package init file 'pyspark/python/pyspark/__init__.py' not found (or not a regular file)
package init file 'lib/__init__.py' not found (or not a regular file)
package init file 'deps/data/__init__.py' not found (or not a regular file)
package init file 'deps/licenses/__init__.py' not found (or not a regular file)
package init file 'deps/conf/__init__.py' not found (or not a regular file)
package init file 'deps/examples/__init__.py' not found (or not a regular file)
reading manifest file 'databricks_connect.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no previously-included files matching '*.py[cod]' found anywhere in distribution
warning: no previously-included files matching '__pycache__' found anywhere in distribution
warning: no previously-included files matching '.DS_Store' found anywhere in distribution
writing manifest file 'databricks_connect.egg-info/SOURCES.txt'
running build_ext
Could not find valid SPARK_HOME while searching ['/', '/build/pyspark', '/build/pyspark', '/']
builder for '/nix/store/pmpx14b0j3zlggdf3kzndf76jdpvqxy1-python3.8-databricks-connect-7.1.0.drv' failed with exit code 255
error: build of '/nix/store/pmpx14b0j3zlggdf3kzndf76jdpvqxy1-python3.8-databricks-connect-7.1.0.drv' failed
```

File: pkgs/development/python-modules/databricks-connect/default.nix (hunk 5 above)
Comment (kfollesdal) [reply pkgs/development/python-modules/databricks-connect/default.nix]: Have now removed py37 constraint.

File: pkgs/development/python-modules/databricks-connect/default.nix (hunk 4 above)
Comment (kfollesdal) [reply pkgs/development/python-modules/databricks-connect/default.nix]: Have now moved sha256 into fetchPypi

File: pkgs/development/python-modules/databricks-connect/default.nix (hunk 3 above)
Comment (kfollesdal) [reply pkgs/development/python-modules/databricks-connect/default.nix:12]: I was thinking that is look cleaner. But have seen now that is is normal to have sha256 hash in fetchPypi. So have now changed to the standard way.

File: lib/licenses.nix (hunk 1 above)
Comment (kfollesdal) [reply lib/licenses.nix]: Was difficult to figure it out in the license. Have now set license to free = false to be on the safe side. At an later point I will try to figure it out.

File: pkgs/development/python-modules/databricks-connect/default.nix (hunk 2 above)
Comment (kfollesdal) [reply pkgs/development/python-modules/databricks-connect/default.nix:20]: Have no take a close look. There are no automatic tests in the package that I can find. 

But after package is installed there is test command `databricks-connect test`, for it to work you have to configure connection to a databricks cluster with `databricks-connect configure`. So user have to run test manually after install. 

Have added a comment about this in nix expression.

**Comment by jonringer** [comment]
is this meant to be used an application or python package? (do you just do `$ databricks-connect` or do you need to do `import databricks-connect` in python?)

**Comment by jonringer** [comment]
according to https://docs.databricks.com/dev-tools/databricks-connect.html it looks like it's meant to be used as a cli tool

**Comment by jonringer** [comment]
if it's only meant to be used from the cli, please use https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonapplication-function as this allows for separation of python environments

File: lib/licenses.nix (hunk 1 above)
Comment (SuperSandro2000) [reply lib/licenses.nix]: Does not seem free to me at first glance.

```
LICENSE

Copyright (2018) Databricks, Inc.

This library (the "Software") may not be used except in connection with the Licensee's use of the Databricks Platform Services pursuant to an Agreement (defined below) between Licensee (defined below) and Databricks, Inc. ("Databricks"). This Software shall be deemed part of the "Subscription Services" under the Agreement, or if the Agreement does not define Subscription Services, then the term in such Agreement that refers to the applicable Databricks Platform Services (as defined below) shall be substituted herein for "Subscription Services." Licensee's use of the Software must comply at all times with any restrictions applicable to the Subscription Services, generally, and must be used in accordance with any applicable documentation. If you have not agreed to an Agreement or otherwise do not agree to these terms, you may not use the Software. This license terminates automatically upon the termination of the Agreement or Licensee's breach of these terms.

Agreement: the agreement between Databricks and Licensee governing the use of the Databricks Platform Services, which shall be, with respect to Databricks, the Databricks Terms of Service located at www.databricks.com/termsofservice, and with respect to Databricks Community Edition, the Community Edition Terms of Service located at www.databricks.com/ce-termsofuse, in each case unless Licensee has entered into a separate written agreement with Databricks governing the use of the applicable Databricks Platform Services.

Databricks Platform Services: the Databricks services or the Databricks Community Edition services, according to where the Software is used.

Licensee: the user of the Software, or, if the Software is being used on behalf of a company, the company.
```

**Comment by kfollesdal** [comment]
You use it as a python package. But have some applications for setup.

Normal user case is: 
1. `$ databricks-connect configure` configure connection to databricks cluster
2. ` $ databricks-connect test` test environment and connection to cluster
3.  Then you use it as a package in your python script. But with the name `pyspark`.
```
from pyspark.sql import SparkSession

spark = SparkSession.builder.getOrCreate()
```

File: lib/licenses.nix (hunk 1 above)
Comment (kfollesdal) [reply lib/licenses.nix]: Yes, that is what I also understand. So keep it as free = false.

File: pkgs/development/python-modules/databricks-connect/default.nix (hunk 2 above)
Comment (catern) [reply pkgs/development/python-modules/databricks-connect/default.nix:20]: I noticed it complaining about a missing SPARK_HOME in that console output, do you need to add a dependency on Java Spark?

File: lib/licenses.nix [hunk 6]
```diff
@@ -436,6 +436,12 @@ lib.mapAttrs (n: v: v // { shortName = n; }) {
   };
 
   # Proprietary binaries; free to redistribute without modification.
+  databricks = {
+    fullName = "Databricks Proprietary License";
+    url = "https://pypi.org/project/databricks-connect";
+    free = false;
+  };
```
Comment (jonringer) [inline lib/licenses.nix:443]: this should be a separate commit:
```
  licenses: add databricks
```

File: maintainers/maintainer-list.nix [hunk 7]
```diff
@@ -4310,6 +4310,12 @@
     githubId = 494012;
     name = "Kevin Cox";
   };
+  kfollesdal = {
+    email = "kfollesdal@gmail.com";
+    github = "kfollesdal";
+    githubId = 546087;
+    name = "Kristoffer K. Føllesdal";
+  };
```
Comment (jonringer) [inline maintainers/maintainer-list.nix:4318]: Please make this into a separate commit:

```
maintainers: add <name>
```

File: pkgs/development/python-modules/databricks-connect/default.nix [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "996a9d0f271f6c7edbd2d85b2efb6ff4e58d15222e80f87ca17fdbf224e17056";
+  };
+
+  sourceRoot = ".";
+
+  propagatedBuildInputs = [ py4j six pkgs.jdk ];
+
+  # User have to run tests command
```
Comment (jonringer) [inline pkgs/development/python-modules/databricks-connect/default.nix:16]:
```suggestion
  # User have to run tests command
```

File: pkgs/development/python-modules/databricks-connect/default.nix [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  };
+
+  sourceRoot = ".";
+
+  propagatedBuildInputs = [ py4j six pkgs.jdk ];
+
+  # User have to run tests command 
+  #   databricks-connect test
+  # manually after install and configuration
+  #   databricks-connect configure
```
Comment (jonringer) [inline pkgs/development/python-modules/databricks-connect/default.nix:19]: probably more succinct
```suggestion
  # requires network access
```

File: pkgs/development/python-modules/databricks-connect/default.nix [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  doCheck = false;
```
Comment (jonringer) [inline pkgs/development/python-modules/databricks-connect/default.nix:20]: If tests are not available, then please use `pythonImportsCheck` to import the most important modules. This isn't as good as unit tests, but will usually give a good indication of run-time errors. Please see [pythonImportsCheck documentation](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pythonimportscheck) for more information.

File: pkgs/development/python-modules/databricks-connect/default.nix [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+  preFixup = ''
+      substituteInPlace "$out/bin/find-spark-home" \
+      --replace find_spark_home.py .find_spark_home.py-wrapped
+  '';
+
+  JAVA_HOME = "${pkgs.jdk}/lib/openjdk";
```
Comment (jonringer) [inline pkgs/development/python-modules/databricks-connect/default.nix:27]: this will only be available during buildtime, if you need it during runtime please do:
```
  makeWrapperArgs = [ "--set JAVA_HOME ${pkgs.jdk}/lib/openjdk" ];
```

This will wrap the entrypoint (`databricks-connect` in this case) with JAVA_HOME. If you need this to be set while using as a python package, then you need to patch the source

File: pkgs/development/python-modules/databricks-connect/default.nix [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+  meta = with lib; {
+    description = "Client for connecting to remote Databricks clusters";
+    homepage = https://pypi.org/project/databricks-connect;
```
Comment (jonringer) [inline pkgs/development/python-modules/databricks-connect/default.nix:31]: rfc 45
```suggestion
    homepage = "https://pypi.org/project/databricks-connect";
```

File: pkgs/top-level/python-packages.nix [hunk 13]
```diff
@@ -653,6 +653,8 @@ in {
 
   databricks-cli = callPackage ../development/python-modules/databricks-cli { };
 
+  databricks-connect = callPackage ../development/python-modules/databricks-connect { };
```
Comment (jonringer) [inline pkgs/top-level/python-packages.nix:656]:
```suggestion
  databricks-connect = callPackage ../development/python-modules/databricks-connect {
    inherit (pkgs) jdk;
  };
```

File: pkgs/development/python-modules/databricks-connect/default.nix (hunk 11 above)
Comment (kfollesdal) [reply pkgs/development/python-modules/databricks-connect/default.nix]: Figured out that JAVA_HOME is set by jdk. So I removed env set of JAVA_HOME.

File: pkgs/top-level/python-packages.nix (hunk 13 above)
Comment (kfollesdal) [reply pkgs/top-level/python-packages.nix]: What is the reason for this change?

I have added your suggestion (see commit 44e9626), but had to edit databricks-connect/default.nix to avoid the following build error:
`error: anonymous function at /home/kristoffer/Desktop/tmp/nixpkgs/pkgs/development/python-modules/databricks-connect/default.nix:1:1 called with unexpected argument 'jdk', at /home/kristoffer/Desktop/tmp/nixpkgs/lib/customisation.nix:69:16`

File: pkgs/development/python-modules/databricks-connect/default.nix (hunk 3 above)
Comment (catern) [reply pkgs/development/python-modules/databricks-connect/default.nix:12]: I mean the sourceRoot = "."

Comment (kfollesdal) [reply pkgs/development/python-modules/databricks-connect/default.nix:12]: Unpacking makes multiple directories. So have to set sourceRoot = "." to tell nix build that we are inside src folder.

File: pkgs/top-level/python-packages.nix (hunk 13 above)
Comment (jonringer) [reply pkgs/top-level/python-packages.nix]: sorry, thought it was what you had before, but it was probably:
```
    jdk = pkgs.openjdk;
```

Comment (jonringer) [reply pkgs/top-level/python-packages.nix]: and the reasoning behind it is that someone could add ` python3Packages.jdk`, and then `callPackage` will bring in that package, rather than the intended `pkgs.openjdk` or what was menat to be referenced.

**Comment by jonringer** [comment]
Please follow [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md) and [manual#submitting-changes-making-patches ](https://nixos.org/nixpkgs/manual/#submitting-changes-making-patches) and squash the fix-up commits.

in this case, you should have two commits:
```
maintainers: ...
pythonPackages.databricks-connect: ...
```

**Comment by kfollesdal** [comment]
Have now clean up the commits, so now there are three commits:
```
licenses: add databricks
maintainers: add kfollesdal
pythonPackages.databricks-connect: init at 7.1.0
```

**Comment by kfollesdal** [comment]
Thanks for all the feedback and help. Have one problem/question left. 

If I make a nix-shell with the following script:
```
with import ./nixpkgs { };

pkgs.mkShell rec {
  buildInputs = [
    python37Packages.databricks-connect
  ];
}
```
then using the python package `pyspark` and running the cli cmd `pyspark` works fine.

But if I install an environment with `nix-env -if` and the file:
```
with import ./nixpkgs {};

python37.withPackages (ps: with ps; [ databricks-connect  ])

```
Then I have to manually set the environment variable `JAVA_HOME` to get `pyspark` python package and `pyspark` cli command to work.

Do you have any clue what the problem can be? I think it can be something with the wrapping.

**Comment by kfollesdal** [comment]
UPDATE
Have now figured out that you do not have to set JAVA_HOME manually after installing `pythonPackages.databricks-connect` if I use the following file when installing with `nix-env`:
```
with import ./nixpkgs {};
buildEnv {
  name = "bricks";
  paths = [
    (python37.withPackages (ps: with ps; [ databricks-connect  ]))
    jdk
  ];
}
```
Then everything works fine. 

But as you can see I have to add `jdk` in the environment. That is strange since it is already added as an dependency in the expression for `pythonPackages.databricks-connect`. Any advice to how I can avoid this?

**Comment by jonringer** [comment]
sorry, I sorted python-packages, can you rebase? :)

**Comment by jonringer** [comment]
oh, i forgot to respond

`JAVA_HOME` currently will only be exposed the builder.

For it to  appear in a nix-shell, you would need to do:
```
mkShell {
  inputsFrom = [ python3Packages.databricks-connect ];
}
```
however, inputsFrom has other implications, it's easier to just a jdk like you did above

**Review by jonringer — APPROVED** [review]
LGTM
shows usage

```
https://github.com/NixOS/nixpkgs/pull/96638
3 packages built:
python27Packages.databricks-connect python37Packages.databricks-connect python38Packages.databricks-connect

```

**Comment by jonringer** [comment]
thanks @kfollesdal for opening your first nixpkgs PR :)

**Comment by kfollesdal** [comment]
Thanks @jonringer, I hope I can contribute more. 

Regarding the `nix-shell` and `nix-env`, I think you misunderstood my question. 

`nix-shell ` works as expected with the following expression: 
```
with import ./nixpkgs { };

pkgs.mkShell {
  buildInputs = [
    python37Packages.databricks-connect
  ];
}
```
But if I install `databricks-connect` with `nix-env` with the 
following expression: 
```
with import ./nixpkgs {};
buildEnv {
  name = "bricks";
  paths = [
    (python37.withPackages (ps: with ps; [ databricks-connect  ]))
  ];
}
```
Then it do not work as expected. But if I add `jdk` to paths, like so: 
```
with import ./nixpkgs {};
buildEnv {
  name = "bricks";
  paths = [
    (python37.withPackages (ps: with ps; [ databricks-connect  ]))
    jdk
  ];
}
```
Then everything works fine as in the `nix-shell` case. I think this is strange since
`jdk` is already added as an dependency to `databricks-connect`. Is it anyway we can fix this 
so it works as expected?

**Comment by jonringer** [comment]
I think your issue is `buildEnv` which I believe just symlink joins the outputs, and will disregard stuff passed to `mkDerivation`

**Comment by kfollesdal** [comment]
Ok, thanks. I will look more into it :-)
